File: app/services/oauth.py
"""
OAuth 2.0 engine — PKCE and device-code flows for providers that don't
take static API keys (Kiro via AWS SSO OIDC, Antigravity via Google Cloud
Code, Gemini Code Assist via Google).

Client ids come from the public registrations used by the official CLIs
(Kiro CLI, Google Cloud Code).

State lives in Redis (shared across workers) with an in-memory fallback.
"""

import base64
import hashlib
import logging
import os
import secrets
import time
from dataclasses import dataclass, field

import httpx

from app.services.oauth_store import oauth_store

logger = logging.getLogger(__name__)


# =============================================================================
# PKCE helpers
# =============================================================================

def pkce_pair() -> tuple[str, str]:
    """Return (verifier, S256 challenge) per RFC 7636."""
    verifier = secrets.token_urlsafe(48)
    digest = hashlib.sha256(verifier.encode()).digest()
    challenge = base64.urlsafe_b64encode(digest).rstrip(b"=").decode()
    return verifier, challenge


def new_state() -> str:
    return secrets.token_urlsafe(24)


# =============================================================================
# Provider specs
# =============================================================================

@dataclass
class OAuthSpec:
    """Redirect/endpoint constants for one OAuth provider."""
    name: str
    flow: str                      # "pkce" | "device_code"
    authorize_url: str
    token_url: str
    client_id: str
    client_secret: str = ""        # public for native apps
    scopes: tuple[str, ...] = ()
    redirect_uri: str = ""
    extra_auth_params: dict = field(default_factory=dict)
    extra_token_params: dict = field(default_factory=dict)
    token_auth_method: str = "client_secret_post"  # Google native apps use body not basic


KIRO_SPEC = OAuthSpec(
    name="kiro",
    flow="device_code",
    authorize_url="https://oidc.us-east-1.amazonaws.com/device_authorization",
    token_url="https://oidc.us-east-1.amazonaws.com/token",
    client_id="",  # filled by client registration at flow start
    client_secret="",
    scopes=(
        "codewhisperer:completions",
        "codewhisperer:analysis",
        "codewhisperer:conversations",
    ),
)

ANTIGRAVITY_SPEC = OAuthSpec(
    name="antigravity",
    flow="pkce",
    authorize_url="https://accounts.google.com/o/oauth2/v2/auth",
    token_url="https://oauth2.googleapis.com/token",
    # Defaults are the public Google Cloud Code client credentials for native
    # apps — Google documents these as distributable in source code. They are
    # still read from env so forks can register their own client.
    # See https://developers.google.com/identity/protocols/oauth2/native-app
    client_id=os.getenv(
        "ANTIGRAVITY_CLIENT_ID",
        "1071006060591-tmhssin2h21lcre235vtolojh4g403ep.apps.googleusercontent.com",
    ),
    client_secret=os.getenv("ANTIGRAVITY_CLIENT_SECRET", ""),
    scopes=(
        "openid",
        "https://www.googleapis.com/auth/cloud-platform",
        "https://www.googleapis.com/auth/userinfo.email",
        "https://www.googleapis.com/auth/userinfo.profile",
        "https://www.googleapis.com/auth/cclog",
        "https://www.googleapis.com/auth/experimentsandconfigs",
    ),
    redirect_uri=os.getenv("ANTIGRAVITY_REDIRECT_URI", "http://localhost:8000/dashboard/oauth/callback"),
    extra_auth_params={"access_type": "offline", "prompt": "consent"},
)

SPECS: dict[str, OAuthSpec] = {s.name: s for s in (KIRO_SPEC, ANTIGRAVITY_SPEC)}


# =============================================================================
# Pending flow state
# =============================================================================


@dataclass
class PendingFlow:
    """One in-flight OAuth handshake waiting for the user to consent."""
    provider: str
    state: str
    verifier: str
    created_at: float = field(default_factory=time.time)
    device_code: str | None = None
    interval: int = 5


_pending: dict[str, PendingFlow] = {}  # state -> PendingFlow
_PENDING_TTL = 600  # 10 minutes


def _gc_pending():
    cutoff = time.time() - _PENDING_TTL
    for state in [s for s, f in _pending.items() if f.created_at < cutoff]:
        _pending.pop(state, None)


# =============================================================================
# PKCE flow (Google)
# =============================================================================

async def start_pkce_flow(spec: OAuthSpec) -> dict:
    """Build the authorization URL + stashed verifier for a PKCE flow."""
    verifier, challenge = pkce_pair()
    state = new_state()

    _gc_pending()
    _pending[state] = PendingFlow(provider=spec.name, state=state, verifier=verifier)

    from urllib.parse import urlencode
    params = {
        "client_id": spec.client_id,
        "redirect_uri": spec.redirect_uri,
        "response_type": "code",
        "scope": " ".join(spec.scopes),
        "code_challenge": challenge,
        "code_challenge_method": "S256",
        "access_type": "offline",
        "prompt": "consent",
        "state": state,
        **spec.extra_auth_params,
    }
    auth_url = f"{spec.authorize_url}?{urlencode(params)}"

    return {
        "provider": spec.name,
        "flow": "pkce",
        "authUrl": auth_url,
        "state": state,
    }


async def complete_pkce_flow(state: str, code: str) -> dict:
    """Exchange the callback code for tokens and store them."""
    flow = _pending.pop(state, None)
    if not flow:
        raise ValueError("OAuth state not found or expired")

    spec = SPECS[flow.provider]

    payload = {
        "grant_type": "authorization_code",
        "code": code,
        "client_id": spec.client_id,
        "client_secret": spec.client_secret,
        "redirect_uri": spec.redirect_uri,
        "code_verifier": flow.verifier,
        **spec.extra_token_params,
    }

    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            spec.token_url,
            data=payload,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

    if resp.status_code != 200:
        raise ValueError(f"token exchange failed: HTTP {resp.status_code} {resp.text[:300]}")

    data = resp.json()
    await _persist_tokens(spec.name, data)

    # Post-exchange: look up the project for Antigravity
    if spec.name == "antigravity":
        await _bootstrap_antigravity_project(data.get("access_token"))

    return {"provider": spec.name, "connected": True}


async def _bootstrap_antigravity_project(access_token: str) -> None:
    """Fetch the Cloud Code project id and email for the antigravity account."""
    try:
        async with httpx.AsyncClient(timeout=30) as client:
            resp = await client.post(
                "https://cloudcode-pa.googleapis.com/v1internal:loadCodeAssist",
                json={"metadata": {"ideType": "ANTIGRAVITY", "pluginType": "GEMINI"}},
                headers={"Authorization": f"Bearer {access_token}"},
            )
            resp.raise_for_status()
            body = resp.json()
            project = body.get("cloudaicompanionProject")
            stored = await oauth_store.get("antigravity") or {}
            stored_meta = stored.get("meta", {})
            stored_meta["project_id"] = project
            stored.setdefault("meta", {}).update(stored_meta)
            await oauth_store.save("antigravity", stored)
    except Exception as e:
        logger.warning(f"Antigravity project bootstrap failed: {e}")


# =============================================================================
# Device Code flow (Kiro / AWS SSO OIDC)
# =============================================================================

async def start_device_flow(spec: OAuthSpec) -> dict:
    """
    Start an AWS SSO OIDC device authorization:
    1. Register a transient public client.
    2. Request a device code.
    3. Return the verification URI + user code + polling info.
    """
    region = "us-east-1"
    oidc_base = f"https://oidc.{region}.amazonaws.com"

    async with httpx.AsyncClient(timeout=30) as client:
        # 1. Register a transient client
        reg_resp = await client.post(
            f"{oidc_base}/client/register",
            json={
                "clientName": f"my-gateway-{spec.name}",
                "clientType": "public",
                "scopes": list(spec.scopes),
                "grantTypes": [
                    "urn:ietf:params:oauth:grant-type:device_code",
                    "refresh_token",
                ],
            },
        )
        if reg_resp.status_code != 200:
            raise ValueError(f"client registration failed: {reg_resp.status_code} {reg_resp.text[:300]}")
        reg = reg_resp.json()
        client_id = reg["clientId"]
        client_secret = reg["clientSecret"]

        # 2. Start device authorization
        dev_resp = await client.post(
            f"{oidc_base}/device_authorization",
            json={
                "clientId": client_id,
                "clientSecret": client_secret,
                "startUrl": f"https://{region}.awsapps.com/start",
            },
        )
        if dev_resp.status_code != 200:
            raise ValueError(f"device authorization failed: {dev_resp.status_code} {dev_resp.text[:300]}")
        dev = dev_resp.json()

    state = new_state()
    _gc_pending()
    _pending[state] = PendingFlow(
        provider=spec.name,
        state=state,
        verifier="",
        device_code=dev["deviceCode"],
        interval=dev.get("interval", 5),
    )

    # Stash the transient client creds so the poller can finish the flow.
    _pending[state].verifier = f"{client_id}:{client_secret}"

    return {
        "provider": spec.name,
        "flow": "device_code",
        "verificationUri": dev["verificationUri"],
        "verificationUriComplete": dev.get("verificationUriComplete", dev["verificationUri"]),
        "userCode": dev["userCode"],
        "expiresIn": dev["expiresIn"],
        "interval": dev.get("interval", 5),
        "state": state,
    }


async def complete_device_flow(state: str) -> dict:
    """Poll OIDC /token until the user authorizes (caller retries on slow_down)."""
    flow = _pending.get(state)
    if not flow or not flow.device_code:
        raise ValueError("OAuth state not found or expired")

    spec = SPECS[flow.provider]
    client_id, client_secret = flow.verifier.split(":", 1)
    oidc_base = "https://oidc.us-east-1.amazonaws.com"

    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            f"{oidc_base}/token",
            json={
                "clientId": client_id,
                "clientSecret": client_secret,
                "deviceCode": flow.device_code,
                "grantType": "urn:ietf:params:oauth:grant-type:device_code",
            },
        )

    if resp.status_code == 400:
        # pending: slow_down / authorization_pending
        err = resp.json()
        code = err.get("error", "")
        if code in ("authorization_pending", "slow_down"):
            raise PendingError(code, interval=flow.interval)
        raise ValueError(f"token poll failed: {err}")

    resp.raise_for_status()
    _pending.pop(state, None)

    # AWS OIDC returns camelCase keys (accessToken/refreshToken/expiresIn)
    raw = resp.json()
    normalized = {
        "access_token": raw.get("accessToken") or raw.get("access_token"),
        "refresh_token": raw.get("refreshToken") or raw.get("refresh_token"),
        "expires_in": raw.get("expiresIn") or raw.get("expires_in", 3600),
        "scope": raw.get("scope", ""),
    }
    return await _persist_tokens(spec.name, normalized)


class PendingError(Exception):
    """Device flow not yet authorized — caller should keep polling."""

    def __init__(self, code: str, interval: int = 5):
        super().__init__(code)
        self.code = code
        self.interval = interval


async def refresh_token_flow(provider: str) -> dict:
    """Refresh an expired OAuth token using its refresh_token."""
    stored = await oauth_store.get(provider)
    if not stored or not stored.get("refresh_token"):
        raise ValueError(f"No refresh token for '{provider}'")

    spec = SPECS[provider]
    payload = {
        "grant_type": "refresh_token",
        "refresh_token": stored["refresh_token"],
        "client_id": spec.client_id,
    }
    if spec.client_secret:
        payload["client_secret"] = spec.client_secret

    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            spec.token_url,
            data=payload,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )

    if resp.status_code != 200:
        raise ValueError(f"refresh failed: HTTP {resp.status_code}")

    data = resp.json()
    data.setdefault("refresh_token", stored["refresh_token"])
    return await _persist_tokens(provider, data)


async def _persist_tokens(provider: str, token_response: dict) -> dict:
    """Normalize a token response and store it."""
    expires_in = int(token_response.get("expires_in", 3600))
    record = {
        "access_token": token_response["access_token"],
        "refresh_token": token_response.get("refresh_token"),
        "expires_at": time.time() + expires_in,
        "scopes": token_response.get("scope", "").split() if token_response.get("scope") else [],
        "meta": {},
    }
    await oauth_store.save(provider, record)
    logger.info(f"OAuth tokens stored for '{provider}' (expires in {expires_in}s)")
    return record


async def get_valid_access_token(provider: str) -> str:
    """Return a fresh access token, refreshing when close to expiry."""
    stored = await oauth_store.get(provider)
    if not stored:
        raise ValueError(f"Provider '{provider}' is not connected - finish OAuth first")

    if await oauth_store.is_expired(provider):
        if stored.get("refresh_token"):
            stored = await refresh_token_flow(provider)
        else:
            raise ValueError(f"Token expired for '{provider}' and it has no refresh token")

    return stored["access_token"]

File: gateway/internal/oauth/oauth.go
// Package oauth implements the OAuth handshake engines: PKCE (Google) and
// AWS SSO OIDC device code (Kiro).
//
// Tokens persist under gw:oauth:{provider} in Redis. The local cache is purely
// a read-through mirror.
package oauth

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

// TokenRecord is the persisted OAuth session.
type TokenRecord struct {
	AccessToken  string         `json:"access_token"`
	RefreshToken string         `json:"refresh_token,omitempty"`
	ExpiresAt    time.Time      `json:"expires_at"`
	Scopes       []string       `json:"scopes"`
	Meta         map[string]any `json:"meta,omitempty"`
}

// PendingFlow holds the state for an in-flight handshake.
type PendingFlow struct {
	Provider   string
	State      string
	Verifier   string // PKCE verifier (also used as client_id:secret for device flow)
	DeviceCode string
	CreatedAt  time.Time
}

// Manager owns flows plus the persisted token store.
type Manager struct {
	rdb       redis.UniversalClient
	http      *http.Client

	mu      sync.Mutex
	pending map[string]*PendingFlow

	local map[string]*TokenRecord // memory mirror of Redis
}

// New returns a manager.
func New(rdb redis.UniversalClient) *Manager {
	return &Manager{
		rdb:     rdb,
		http:    &http.Client{Timeout: 30 * time.Second},
		pending: map[string]*PendingFlow{},
		local:   map[string]*TokenRecord{},
	}
}

// ------------------------- PKCE (Google) -------------------------

var pkceAlphabet = []rune("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~")

func randomString(n int) string {
	buf := make([]byte, n)
	_, _ = rand.Read(buf)
	out := make([]rune, n)
	for i, b := range buf {
		out[i] = pkceAlphabet[int(b)%len(pkceAlphabet)]
	}
	return string(out)
}

// sha256Sum is sha256(data).
func sha256Sum(b []byte) [32]byte { return sha256.Sum256(b) }

// base64URLEncode is base64-url without padding (RFC 4648 §5).
func base64URLEncode(b []byte) string {
	return base64.RawURLEncoding.EncodeToString(b)
}

type PKCESpec struct {
	Name          string
	ClientID      string
	ClientSecret  string
	AuthorizeURL  string
	TokenURL      string
	RedirectURI   string
	Scopes        []string
}

// StartPKCE builds an authorization URL and stores the pending verifier.
func (m *Manager) StartPKCE(spec PKCESpec) (string, string, error) {
	verifier := randomString(64)
	sum := sha256Sum([]byte(verifier))
	challenge := base64URLEncode(sum[:])

	state := randomString(24)
	m.mu.Lock()
	m.pending[state] = &PendingFlow{
		Provider: spec.Name,
		State:    state,
		Verifier: verifier,
		CreatedAt: time.Now(),
	}
	m.mu.Unlock()

	q := url.Values{}
	q.Set("client_id", spec.ClientID)
	q.Set("redirect_uri", spec.RedirectURI)
	q.Set("response_type", "code")
	q.Set("scope", strings.Join(spec.Scopes, " "))
	q.Set("code_challenge", challenge)
	q.Set("code_challenge_method", "S256")
	q.Set("access_type", "offline")
	q.Set("prompt", "consent")
	q.Set("state", state)

	return spec.AuthorizeURL + "?" + q.Encode(), state, nil
}

// CompletePKCE exchanges a callback code for tokens and stores them.
func (m *Manager) CompletePKCE(state, code string, spec PKCESpec) (*TokenRecord, error) {
	m.mu.Lock()
	flow, ok := m.pending[state]
	m.mu.Unlock()
	if !ok {
		return nil, errors.New("unknown or expired OAuth state")
	}

	form := url.Values{}
	form.Set("grant_type", "authorization_code")
	form.Set("code", code)
	form.Set("client_id", spec.ClientID)
	form.Set("client_secret", spec.ClientSecret)
	form.Set("redirect_uri", spec.RedirectURI)
	form.Set("code_verifier", flow.Verifier)

	req, err := http.NewRequest(http.MethodPost, spec.TokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")

	resp, err := m.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return nil, fmt.Errorf("token exchange failed: %d %s", resp.StatusCode, string(body))
	}
	var tr map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&tr); err != nil {
		return nil, err
	}

	rec := parseTokenResponse(tr)
	if rec.AccessToken == "" {
		return nil, errors.New("token response missing access_token")
	}
	m.mu.Lock()
	delete(m.pending, state)
	m.mu.Unlock()

	if err := m.save(context.Background(), spec.Name, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// ------------------------- Device code (AWS OIDC) -------------------------

type DeviceCodeSpec struct {
	Provider        string
	OIDCBase        string // https://oidc.us-east-1.amazonaws.com
	Scopes          []string
}

// StartDeviceCode registers a transient OIDC client and starts a device flow.
func (m *Manager) StartDeviceCode(spec DeviceCodeSpec) (map[string]any, error) {
	registerBody := map[string]any{
		"clientName": "my-gateway-" + spec.Provider,
		"clientType": "public",
		"scopes":     spec.Scopes,
		"grantTypes": []string{
			"urn:ietf:params:oauth:grant-type:device_code",
			"refresh_token",
		},
	}
	reg, err := m.postJSON(spec.OIDCBase+"/client/register", registerBody)
	if err != nil {
		return nil, err
	}
	clientID, _ := reg["clientId"].(string)
	clientSecret, _ := reg["clientSecret"].(string)
	if clientID == "" || clientSecret == "" {
		return nil, errors.New("OIDC registration missing clientId/clientSecret")
	}

	dev, err := m.postJSON(spec.OIDCBase+"/device_authorization", map[string]any{
		"clientId":     clientID,
		"clientSecret": clientSecret,
		"startUrl":     "https://device.sso.us-east-1.amazonaws.com/",
	})
	if err != nil {
		return nil, err
	}

	deviceCode, _ := dev["deviceCode"].(string)
	userCode, _ := dev["userCode"].(string)
	verificationURI, _ := dev["verificationUri"].(string)
	verificationURIComplete, _ := dev["verificationUriComplete"].(string)
	expiresIn, _ := dev["expiresIn"].(float64)
	interval, _ := dev["interval"].(float64)
	if interval <= 0 {
		interval = 5
	}

	state := randomString(24)
	m.mu.Lock()
	m.pending[state] = &PendingFlow{
		Provider:   spec.Provider,
		State:      state,
		Verifier:   clientID + ":" + clientSecret,
		DeviceCode: deviceCode,
		CreatedAt:  time.Now(),
	}
	m.mu.Unlock()

	return map[string]any{
		"provider":                 spec.Provider,
		"flow":                     "device_code",
		"verificationUri":          verificationURI,
		"verificationUriComplete":  verificationURIComplete,
		"userCode":                 userCode,
		"expiresIn":                int(expiresIn),
		"interval":                 int(interval),
		"state":                    state,
	}, nil
}

// PollDeviceCode checks one device-code flow; 202 while pending, error later.
func (m *Manager) PollDeviceCode(state string) (*TokenRecord, bool, error) {
	m.mu.Lock()
	flow, ok := m.pending[state]
	m.mu.Unlock()
	if !ok {
		return nil, false, errors.New("unknown OAuth state")
	}

	parts := strings.SplitN(flow.Verifier, ":", 2)
	if len(parts) != 2 {
		return nil, false, errors.New("malformed client credentials")
	}

	tokenBody := map[string]any{
		"clientId":     parts[0],
		"clientSecret": parts[1],
		"deviceCode":   flow.DeviceCode,
		"grantType":    "urn:ietf:params:oauth:grant-type:device_code",
	}
	tr, err := m.postJSON("https://oidc.us-east-1.amazonaws.com/token", tokenBody)
	if err != nil {
		// AWS returns 400 with `authorization_pending` while waiting
		msg := err.Error()
		if strings.Contains(msg, "authorization_pending") || strings.Contains(msg, "slow_down") {
			return nil, false, nil
		}
		return nil, false, err
	}

	rec := parseTokenResponse(tr)
	if rec.AccessToken == "" {
		return nil, false, errors.New("token response missing access token")
	}

	m.mu.Lock()
	delete(m.pending, state)
	m.mu.Unlock()

	if err := m.save(context.Background(), flow.Provider, rec); err != nil {
		return nil, false, err
	}
	return rec, true, nil
}

// ------------------------- Token storage -------------------------

func (m *Manager) key(provider string) string { return "gw:oauth:" + provider }

// Save persists a token.
func (m *Manager) save(ctx context.Context, provider string, rec *TokenRecord) error {
	buf, err := json.Marshal(rec)
	if err != nil {
		return err
	}
	if err := m.rdb.Set(ctx, m.key(provider), buf, 0).Err(); err != nil {
		return err
	}
	m.mu.Lock()
	m.local[provider] = rec
	m.mu.Unlock()
	return nil
}

// Get returns the stored record, or nil.
func (m *Manager) Get(provider string) *TokenRecord {
	m.mu.Lock()
	if rec, ok := m.local[provider]; ok {
		m.mu.Unlock()
		return rec
	}
	m.mu.Unlock()

	raw, err := m.rdb.Get(context.Background(), m.key(provider)).Result()
	if err != nil {
		return nil
	}
	var rec TokenRecord
	if json.Unmarshal([]byte(raw), &rec) != nil {
		return nil
	}
	m.mu.Lock()
	m.local[provider] = &rec
	m.mu.Unlock()
	return &rec
}

// AccessToken returns the current access token; refreshes if near expiry.
func (m *Manager) AccessToken(provider string) string {
	rec := m.Get(provider)
	if rec == nil {
		return ""
	}
	if time.Until(rec.ExpiresAt) < 60*time.Second && rec.RefreshToken != "" {
		if refreshed, err := m.refresh(provider, rec); err == nil {
			rec = refreshed
		}
	}
	return rec.AccessToken
}

// ProjectID returns a provider-specific extra (e.g. antigravity's cloud project).
func (m *Manager) ProjectID(provider string) string {
	rec := m.Get(provider)
	if rec == nil || rec.Meta == nil {
		return ""
	}
	if v, ok := rec.Meta["project_id"].(string); ok {
		return v
	}
	return ""
}

// ConnectedProviders lists providers with live tokens.
func (m *Manager) ConnectedProviders() map[string]map[string]any {
	out := map[string]map[string]any{}
	m.mu.Lock()
	defer m.mu.Unlock()
	for name, rec := range m.local {
		out[name] = map[string]any{
			"connected":   time.Now().Before(rec.ExpiresAt),
			"expires_in":  int(time.Until(rec.ExpiresAt).Seconds()),
		}
	}
	// Merge any tokens that are only in Redis (multi-worker restart)
	if keys, err := m.rdb.Keys(context.Background(), "gw:oauth:*").Result(); err == nil {
		for _, key := range keys {
			name := strings.TrimPrefix(key, "gw:oauth:")
			if _, seen := out[name]; seen {
				continue
			}
			if rec := m.Get(name); rec != nil {
				out[name] = map[string]any{
					"connected":  time.Now().Before(rec.ExpiresAt),
					"expires_in": int(time.Until(rec.ExpiresAt).Seconds()),
				}
			}
		}
	}
	return out
}

// Delete removes a provider's token.
func (m *Manager) Delete(provider string) bool {
	m.mu.Lock()
	delete(m.local, provider)
	m.mu.Unlock()
	n, _ := m.rdb.Del(context.Background(), m.key(provider)).Result()
	return n > 0
}

func (m *Manager) refresh(provider string, rec *TokenRecord) (*TokenRecord, error) {
	// Generic refresh — calls the same token endpoint with refresh_token grant.
	// The caller knows the right token URL; for our two providers they're both
	// in the spec. To stay simple we store the token URL in the record's Meta.
	tokenURL, _ := rec.Meta["token_url"].(string)
	if tokenURL == "" {
		return nil, errors.New("no token URL stored for refresh")
	}
	form := url.Values{}
	form.Set("grant_type", "refresh_token")
	form.Set("refresh_token", rec.RefreshToken)
	form.Set("client_id", fmt.Sprint(rec.Meta["client_id"]))

	req, err := http.NewRequest(http.MethodPost, tokenURL, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := m.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("refresh failed: %d", resp.StatusCode)
	}
	var out map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, err
	}
	fresh := parseTokenResponse(out)
	if fresh.RefreshToken == "" {
		fresh.RefreshToken = rec.RefreshToken
	}
	if err := m.save(context.Background(), provider, fresh); err != nil {
		return nil, err
	}
	return fresh, nil
}

func parseTokenResponse(m map[string]any) *TokenRecord {
	access, _ := m["access_token"].(string)
	if access == "" {
		access, _ = m["accessToken"].(string)
	}
	rec := &TokenRecord{
		AccessToken:  access,
		RefreshToken: firstStr(m, "refresh_token", "refreshToken"),
		Scopes:       splitScopes(firstStr(m, "scope", "scopes")),
		Meta:         map[string]any{},
	}
	if v, ok := m["expires_in"].(float64); ok {
		rec.ExpiresAt = time.Now().Add(time.Duration(v) * time.Second)
	} else if v := m["expiresIn"]; v != nil {
		if f, ok := v.(float64); ok {
			rec.ExpiresAt = time.Now().Add(time.Duration(f) * time.Second)
		}
	}
	if rec.RefreshToken == "" {
		rec.RefreshToken = firstStr(m, "refreshToken", "refresh_token")
	}
	return rec
}

func firstStr(m map[string]any, keys ...string) string {
	for _, k := range keys {
		if v, ok := m[k].(string); ok && v != "" {
			return v
		}
	}
	return ""
}

func splitScopes(v string) []string {
	if v == "" {
		return nil
	}
	return strings.Fields(v)
}

func (m *Manager) postJSON(url string, body any) (map[string]any, error) {
	buf, err := json.Marshal(body)
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodPost, url, io.NopCloser(bytesReader(buf)))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := m.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	raw, _ := io.ReadAll(io.LimitReader(resp.Body, 64<<10))
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(raw))
	}
	var out map[string]any
	if err := json.Unmarshal(raw, &out); err != nil {
		return nil, err
	}
	return out, nil
}

// bytesReader wraps []byte into io.Reader without extra imports.
func bytesReader(b []byte) io.Reader { return strings.NewReader(string(b)) }
